File: 658.find-k-closest-elements.py
#
# @lc app=leetcode id=658 lang=python3
#
# [658] Find K Closest Elements
#

# @lc code=start
class Solution:
    def findClosestElements(self, arr: list[int], k: int, x: int) -> list[int]:
        #use the eclusion method
        # left, right = 0, len(arr) - 1
        # if right - left == 0:
        #     return arr
        # while (right - left + 1) > k: 
        #     if abs(arr[left] - x) <= abs(arr[right] - x):
        #         right -= 1
        #     else:
        #         left += 1
        # return arr[left:right + 1]
        #

        #todo use the binary search method
        left, right = 0, len(arr) - k

        while left < right:
            mid = (left + right) // 2
            if x - arr[mid] > arr[mid + k] - x:
                left = mid + 1
            else:
                right = mid
        return arr[left:left + k]

# @lc code=end

File: 658.find-k-closest-elements.java
/*
 * @lc app=leetcode id=658 lang=java
 *
 * [658] Find K Closest Elements
 */

// @lc code=start
import java.util.*;
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        int left = 0;
        int right = arr.length - 1;
        int remove = arr.length - k;

        while(remove > 0){
            if(x - arr[left] <= arr[right] - x){
                right--;
            }else{
                left++;
            }
            remove--;
        }
        List<Integer> ans = new ArrayList<>();
        for(int i = left; i < left + k; i++){
            ans.add(arr[i]);
        }
        return ans; 
    }
}
// @lc code=end

